C7/PCXToGodot.cs

```diff
@@ -5,65 +5,102 @@
+	// However, this will become a problem if (when?) we move towards parallelization
+	// But for now, it's an improvement
+	public static int[] BufferData = new int[3840 * 2160];
+	public static int[] ColorData = new int[256];
+	public static int[] AlphaData = new int[256];
+
+	public static void loadPalette(byte[,] palette, bool shadows = false) {
+		int Red, Green, Blue;
+		int Alpha = 255 << 24;
+
+		for (int i = 0; i < 256; i++) {
+			Red = palette[i, 0];
+			Green = palette[i, 1] << 8;
+			Blue = palette[i, 2] << 16;
+			ColorData[i] = Red + Green + Blue + Alpha;
+		}
+
+		for (int i = CIV3_TRANSPARENCY_START; i < 256; i++) {
+			ColorData[i] &= 0x00ffffff;
+		}
+
+		if (shadows) {
+			for (int i = 240; i < 256; i++) {
+				ColorData[i] = ((255 - i) * 16) << 24;
+			}
+		}
+	}
+
+	public static void loadAlphaPalette(byte[,] palette) {
+		for (int i = 0; i < 256; i++) {
+			// Assumption based on menuButtonsAlpha.pcx: The palette in the alpha PCX always has the same red, green, and blue values (i.e. is grayscale).
+			// Examining it with breakpoints in my Java code, it appears it starts at 255, 255, 255, and goes down one at a time.  But this code
+			// doesn't assume that, it only assumes the grayscale aspect.  In theory, this should work for any transparency, 0 to 255.
+			AlphaData[i] = palette[i, 0] << 24;
+			ColorData[i] = ColorData[i] &= 0x00ffffff;
+		}
+	}
+
+	public static Image getImageFromBufferData(int width, int height) {
+		Image image = new Image();
+		var Data = new byte[4 * width * height];
+		Buffer.BlockCopy(BufferData, 0, Data, 0, 4 * width * height);
+		image.CreateFromData(width, height, false, Image.Format.Rgba8, Data);
+		return image;
 	}
 
 	/**
 	 * This method is for cases where we want to use components of multiple PCXs in a texture, such as for the popup background.
 	 **/
 	public static Image getImageFromPCX(Pcx pcx, int leftStart, int topStart, int croppedWidth, int croppedHeight) {
-		Image image = new Image();
-		image.Create(croppedWidth, croppedHeight, false, Image.Format.Rgba8);
-		image.Lock();
-		for (int y = topStart; y < topStart + croppedHeight; y++)
-		{
-			for (int x = leftStart; x < leftStart + croppedWidth; x++)
-			{
-				byte red = pcx.Palette[pcx.ColorIndexAt(x, y), 0];
-				byte green = pcx.Palette[pcx.ColorIndexAt(x, y), 1];
-				byte blue = pcx.Palette[pcx.ColorIndexAt(x, y), 2];
-				byte alpha = pcx.ColorIndexAt(x, y) >= CIV3_TRANSPARENCY_START ? (byte)0 : (byte)255;
-				image.SetPixel(x - leftStart, y - topStart, Color.Color8(red, green, blue, alpha));
+		loadPalette(pcx.Palette);
+
+		int yEnd = topStart + croppedHeight;
+		int xEnd = leftStart + croppedWidth;
+		int Index;
+		int DataIndex = 0;
+
+		for (int y = topStart; y < yEnd; y++) {
+			Index = y * pcx.Width + leftStart;
+			for (int x = leftStart; x < xEnd; x++) {
+				BufferData[DataIndex++] = ColorData[pcx.ColorIndices[Index++]];
```

Comment:
Minor, but why not just move Index = to inside the x loop, and calculate it once there?

```
for (int x = leftStart; x < xEnd; x++) {
    Index = y * pcx.Width + leftStart + x;
}
```

It's probably a readability versus math overhead thing.  TBH, I'd just as soon leave one call to ColorIndexAt, which is designed for this purpose:

```
		for (int y = topStart; y < yEnd; y++) {
			for (int x = leftStart; x < xEnd; x++) {
				BufferData[DataIndex++] = ColorData[pcx.ColorIndexAt(x, y)];
```

Yes, there's slightly technically more multiplication, but can the difference be measured as noticeable?  There's still a 75% decrease with the calculations versus the old method as well.

If there _is_ a significant difference, it may make sense to leave it as-is, but I've learned that it takes an awful lot of multiplication to matter on a modern CPU, so I'm inclined to go with the more readable option without data on why it matters in that case.

Reply:
Looks like the performance difference here is even more minor than the main Pcx loop changes, so this can be reverted for readability's sake.